File: code/seungwookim99/week2/prog_43163.cpp
// C++ 풀이
#include <string>
#include <iostream>
#include <vector>
#define MAX 1000000000

using namespace std;

int answer = MAX;

int different_alphabets_num(string a, string b) {
    int cnt = 0;
    for(int i = 0 ; i < a.length() ; i++) {
        if (a[i] != b[i]) cnt++;
    }
    return cnt;
}

void dfs(int cnt, string begin, string target, vector<string> words, vector<bool> visited) {
    if (begin.compare(target) == 0) {
        answer = min(answer, cnt);
        return;
    }
    for(int i = 0 ; i < visited.size() ; i++) {
        if (visited[i]) continue;
        if (different_alphabets_num(words[i], begin) == 1) {
            visited[i] = true;
            dfs(cnt + 1, words[i], target, words, visited);
            visited[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    vector<bool> visited(words.size(), false);
    dfs(0, begin, target, words, visited);
    if (answer == MAX) answer = 0;
    return answer;
}

File: code/kauthenticity/week2/prog_43163.py
from collections import deque

MAX_NUM = 1000000000000

def isChangable(a, b) : 
    diff = 0
    
    for i in range(len(a)) : 
        if a[i] != b[i] : 
            diff += 1
        
        if diff == 2 :
            return False
    
    return True

def solution(begin, target, words):
    answer = MAX_NUM
    visited = {}
    
    for word in words : 
        visited[word] = False
        
    queue = deque([(begin, 0)])
    
    while queue : 
        cur, cnt = queue.popleft()
        
        if cur == target : 
            answer = cnt
            break
            
        for word in words : 
            if visited[word]:
                continue

            if isChangable(cur, word):
                queue.append((word, cnt + 1))
                visited[word] = True

    return 0 if answer == MAX_NUM else answer